omnigibson/transition_rules.py

```diff
@@ -296,358 +710,453 @@ def transition(self, individual_objects, group_objects):
-        self._current_steps = 1
-        return True
+    @classproperty
+    def conditions(cls):
+        # Only heated objects are valid
+        return [StateCondition(filter_name="meltable", state=Temperature, val=m.MELTING_TEMPERATURE, op=operator.ge)]
 
-    def transition(self, container_obj, *contained_objs):
+    @classmethod
+    def transition(cls, individual_objects, group_objects):
         t_results = TransitionResults()
 
-        # Create a new object to be added.
-        all_pos, all_orn = [], []
-        for contained_obj in contained_objs:
-            pos, orn = contained_obj.get_position_orientation()
-            all_pos.append(pos)
-            all_orn.append(orn)
-
-        category = self.obj_attrs.category
-        model = self.obj_attrs.model
-        name = f"{self.obj_attrs.name}_{self._counter}"
-        self._counter += 1
-        scale = self.obj_attrs.scale
-
-        model_root_path = f"{gm.DATASET_PATH}/objects/{category}/{model}"
-        usd_path = f"{model_root_path}/usd/{model}.usd"
-
-        final_obj = DatasetObject(
-            prim_path=f"/World/{name}",
-            usd_path=usd_path,
-            category=category,
-            name=f"{name}",
-            scale=scale)
-
-        final_obj_attrs = ObjectAttrs(
-            obj=final_obj, pos=np.mean(all_pos, axis=0), orn=np.mean(all_orn, axis=0))
-        t_results.add.append(final_obj_attrs)
-
-        # Delete all objects inside the container.
-        for contained_obj in contained_objs:
-            t_results.remove.append(contained_obj)
-
-        # Turn off the container, otherwise things would turn into garbage.
-        if ToggledOn in container_obj.states:
-            container_obj.states[ToggledOn].set_value(False)
-        print(f"Applied {ContainerRule.__name__} to {container_obj}")
+        # Convert the meltable object into its melted substance
+        meltable_obj = individual_objects["meltable"]
+        system = get_system(f"melted_{meltable_obj.category}")
+        system.generate_particles_from_link(meltable_obj, meltable_obj.root_link, use_visual_meshes=False)
+
+        # Delete original object from stage.
+        t_results.remove.append(meltable_obj)
+
         return t_results
 
 
-class ContainerGarbageRule(BaseTransitionRule):
+class MixingRule(BaseTransitionRule):
     """
-    Rule to apply to a container to turn what remain inside into garbage.
-
-    This rule is used as a catch-all rule for containers to turn objects inside
-    the container that did not match any other legitimate rules all into a
-    single garbage object.
+    Transition rule to apply to mixing objects together
     """
+    # Maps recipe name to recipe information
+    # NOTE: By defining this class variable directly here, all subclasses SHARE this same dictionary!
+    # This is intentional because there may be multiple valid mixing methods,
+    # and all of them should share the same global recipe pool
+    _RECIPES = dict()
```

Comment:
So are `BlenderRule` and `MixingWandRule` sharing the same `_RECIPES`? Would we lose some efficiency compared to keeping two separate lists of recipes?

Reply:
make _RECIPES initialized separately since recipe list is different between mixing tool and blending rules